src/gridtools/clang/storage_dsl.hpp

```diff
@@ -83,6 +83,390 @@ struct storage {
+};
+
+/**
+ * @brief Dummy 2-dimensional storage
+ * @ingroup gridtools_clang
+ */
+struct storage_ij : public storage {
+  storage_ij();
+
+  storage_ij& operator=(const storage&);
+  storage_ij& operator+=(const storage&);
+  storage_ij& operator-=(const storage&);
+  storage_ij& operator/=(const storage&);
+  storage_ij& operator*=(const storage&);
+
+  storage_ij& operator()(int, int);
+
+  /**
+   * @name 1D access
+   * @{
+   */
+  storage_ij& operator()(dimension);
+
+  storage_ij& operator()(direction);
+  storage_ij& operator()(offset);
+  /** @} */
+
+  /**
+   * @name 2D access
+   * @{
+   */
+
+  storage_ij& operator()(dimension, dimension);
+
+  storage_ij& operator()(dimension, direction);
+  storage_ij& operator()(direction, dimension);
+  storage_ij& operator()(direction, direction);
+  /** @} */
+
+  /**
+   * @name 3D access
+   * @{
+   */
+  storage_ij& operator()(dimension, dimension, dimension);
+
+  storage_ij& operator()(direction, direction, direction);
+  storage_ij& operator()(dimension, direction, direction);
+  storage_ij& operator()(direction, dimension, direction);
+  storage_ij& operator()(direction, direction, dimension);
+  storage_ij& operator()(dimension, dimension, direction);
+  storage_ij& operator()(dimension, direction, dimension);
+  storage_ij& operator()(direction, dimension, dimension);
+  /** @} */
+
+  operator double() const;
+};
+
+/**
+ * @brief Dummy 2-dimensional storage
+ * @ingroup gridtools_clang
+ */
+struct storage_ik : public storage {
+  storage_ik();
+
+  storage_ik& operator=(const storage&);
+  storage_ik& operator+=(const storage&);
+  storage_ik& operator-=(const storage&);
+  storage_ik& operator/=(const storage&);
+  storage_ik& operator*=(const storage&);
+
+  storage_ik& operator()(int, int);
+
+  /**
+   * @name 1D access
+   * @{
+   */
+  storage_ik& operator()(dimension);
+
+  storage_ik& operator()(direction);
+  storage_ik& operator()(offset);
+  /** @} */
+
+  /**
+   * @name 2D access
+   * @{
+   */
+
+  storage_ik& operator()(dimension, dimension);
+
+  storage_ik& operator()(dimension, direction);
+  storage_ik& operator()(direction, dimension);
+  storage_ik& operator()(direction, direction);
+  /** @} */
+
+  /**
+ * @name 3D access
+ * @{
+ */
+  storage_ik& operator()(dimension, dimension, dimension);
+
+  storage_ik& operator()(direction, direction, direction);
+  storage_ik& operator()(dimension, direction, direction);
+  storage_ik& operator()(direction, dimension, direction);
+  storage_ik& operator()(direction, direction, dimension);
+  storage_ik& operator()(dimension, dimension, direction);
+  storage_ik& operator()(dimension, direction, dimension);
+  storage_ik& operator()(direction, dimension, dimension);
+  /** @} */
+
+  operator double() const;
+};
+
+/**
+ * @brief Dummy 2-dimensional storage
+ * @ingroup gridtools_clang
+ */
+struct storage_jk : public storage {
+  storage_jk();
+
+  storage_jk& operator=(const storage&);

```

Comment:
all these classes have very repetitive code, wouldnt a macro autogenerating the class content help?

Reply:
done